In [94]:
# COMP 247 - Project - Group 4
import pandas as pd
import numpy as np

Reading dataset

In [95]:
dataset = pd.read_csv("Bicycle_Thefts.csv", header=0, low_memory=False)

In [96]:
df = pd.DataFrame(dataset)
df

X             Y  OBJECTID event_unique_id  \
0     -8.850630e+06  5.411196e+06     17744  GO-20179016397   
1     -8.850439e+06  5.412149e+06     17759  GO-20172033056   
2     -8.851203e+06  5.411169e+06     17906  GO-20189030822   
3     -8.851203e+06  5.411169e+06     17962   GO-2015804467   
4     -8.851160e+06  5.411032e+06     17963  GO-20159002781   
...             ...           ...       ...             ...   
25564 -8.818235e+06  5.435746e+06      9361   GO-2015543181   
25565 -8.818471e+06  5.436317e+06     11318  GO-20169004589   
25566 -8.820513e+06  5.433291e+06     11462  GO-20169005434   
25567 -8.816571e+06  5.433957e+06     11695  GO-20161170896   
25568 -8.820661e+06  5.433802e+06     11883  GO-20169007653   

             Primary_Offence         Occurrence_Date  Occurrence_Year  \
0                THEFT UNDER  2017/10/03 04:00:00+00             2017   
1      THEFT UNDER - BICYCLE  2017/11/08 05:00:00+00             2017   
2      THEFT UNDER - BICYCLE  2018/09/14 04:00:00+00             2018   
3                THEFT UNDER  2015/05/07 04:00:00+00             2015   
4                THEFT UNDER  2015/05/16 04:00:00+00             2015   
...                      ...                     ...              ...   
25564         MISCHIEF UNDER  2015/04/01 04:00:00+00             2015   
25565            THEFT UNDER  2016/05/16 04:00:00+00             2016   
25566            THEFT UNDER  2016/06/04 04:00:00+00             2016   
25567            THEFT UNDER  2016/07/04 04:00:00+00             2016   
25568  THEFT UNDER - BICYCLE  2016/07/22 04:00:00+00             2016   

      Occurrence_Month Occurrence_DayOfWeek  Occurrence_DayOfMonth  ...  \
0              October              Tuesday                      3  ...   
1             November            Wednesday                      8  ...   
2            September               Friday                     14  ...   
3                  May             Thursday                      7  ...   
4                  May             Saturday                     16  ...   
...                ...                  ...                    ...  ...   
25564            April            Wednesday                      1  ...   
25565              May               Monday                     16  ...   
25566             June             Saturday                      4  ...   
25567             July               Monday                      4  ...   
25568             July               Friday                     22  ...   

          Bike_Make       Bike_Model Bike_Type  Bike_Speed Bike_Colour  \
0                GI         ESCAPE 2        OT           7         BLK   
1      UNKNOWN MAKE              NaN        TO           1         BLK   
2                OT       CROSSTRAIL        MT          24         BLK   
3                GT              NaN        TO          10      BLKDGR   
4                GI              NaN        MT           6         RED   
...             ...              ...       ...         ...         ...   
25564  UNKNOWN MAKE     BMX WILD MAN        RG           0         SIL   
25565            SC              NaN        OT          14         NaN   
25566            SC           ANTRIM        MT          24         WHI   
25567  UNKNOWN MAKE              NaN        SC           1         NaN   
25568            SU  ASCENT MOUNTAIN        MT          21         ONG   

      Cost_of_Bike     Status  Longitude   Latitude ObjectId2  
0            700.0     STOLEN -79.506560  43.648427         1  
1           1100.0  RECOVERED -79.504849  43.654623         2  
2            904.0     STOLEN -79.511709  43.648253         3  
3            400.0     STOLEN -79.511709  43.648253         4  
4            600.0     STOLEN -79.511327  43.647364         5  
...            ...        ...        ...        ...       ...  
25564        600.0     STOLEN -79.215553  43.807798     25565  
25565        900.0     STOLEN -79.217670  43.811497     25566  
25566        700.0

Dataset features type

In [97]:
df.dtypes

X                        float64
Y                        float64
OBJECTID                   int64
event_unique_id           object
Primary_Offence           object
Occurrence_Date           object
Occurrence_Year            int64
Occurrence_Month          object
Occurrence_DayOfWeek      object
Occurrence_DayOfMonth      int64
Occurrence_DayOfYear       int64
Occurrence_Hour            int64
Report_Date               object
Report_Year                int64
Report_Month              object
Report_DayOfWeek          object
Report_DayOfMonth          int64
Report_DayOfYear           int64
Report_Hour                int64
Division                  object
City                      object
Hood_ID                   object
NeighbourhoodName         object
Location_Type             object
Premises_Type             object
Bike_Make                 object
Bike_Model                object
Bike_Type                 object
Bike_Speed                 int64
Bike_Colour               object
Cost_of_Bi

In [98]:
# Keys
df.keys()

Index(['X', 'Y', 'OBJECTID', 'event_unique_id', 'Primary_Offence',
       'Occurrence_Date', 'Occurrence_Year', 'Occurrence_Month',
       'Occurrence_DayOfWeek', 'Occurrence_DayOfMonth', 'Occurrence_DayOfYear',
       'Occurrence_Hour', 'Report_Date', 'Report_Year', 'Report_Month',
       'Report_DayOfWeek', 'Report_DayOfMonth', 'Report_DayOfYear',
       'Report_Hour', 'Division', 'City', 'Hood_ID', 'NeighbourhoodName',
       'Location_Type', 'Premises_Type', 'Bike_Make', 'Bike_Model',
       'Bike_Type', 'Bike_Speed', 'Bike_Colour', 'Cost_of_Bike', 'Status',
       'Longitude', 'Latitude', 'ObjectId2'],
      dtype='object')

Select the only important keys: Month, Area, Cost, and Status of stolen features

In [99]:
df = df[['Occurrence_Month', 'Division', 'Hood_ID', 'Cost_of_Bike', 'Status']]

In [100]:
# Rename some column just for easy co
df=df.rename(columns={'Occurrence_Month':'Month', 'Division':'Div', "Hood_ID" : "Hood", "Cost_of_Bike" : "Cost"})
df.head(5)

Month  Div Hood    Cost     Status
0    October  D22   15   700.0     STOLEN
1   November  D22   15  1100.0  RECOVERED
2  September  D22   15   904.0     STOLEN
3        May  D22   15   400.0     STOLEN
4        May  D22   15   600.0     STOLEN

Evaluate Month feature

In [101]:
df.Month.value_counts()

July         4002
August       3680
June         3532
September    3265
May          2659
October      2492
April        1553
November     1467
March         876
December      834
January       640
February      569
Name: Month, dtype: int64

Evaluate Neigborhood feature

In [102]:
df.Hood.value_counts()

77     2576
76     2112
75     1676
82      992
95      947
       ... 
116      12
3        10
112       9
4         7
29        6
Name: Hood, Length: 141, dtype: int64

Evaluate Division feature

In [103]:
df.Div.value_counts()

D14    4580
D52    4389
D51    4100
D53    2287
D55    1959
D11    1808
D32    1235
D22    1116
D13     966
D54     734
D41     465
D31     421
D43     417
D33     392
D12     250
D23     223
D42     218
NSA       9
Name: Div, dtype: int64

We have NSA which is unknown location, so we assign '000' to this Value

In [104]:
# replace NSA in Division
df.replace("NSA", "D00", inplace=True)
df.Div.value_counts()

D14    4580
D52    4389
D51    4100
D53    2287
D55    1959
D11    1808
D32    1235
D22    1116
D13     966
D54     734
D41     465
D31     421
D43     417
D33     392
D12     250
D23     223
D42     218
000       9
Name: Div, dtype: int64

Evaluate Status feature

In [105]:
df.Status.value_counts()

STOLEN       24807
UNKNOWN        454
RECOVERED      308
Name: Status, dtype: int64

We have 454 case UNKNOWN that means Lost bike (stolen or not), so, we assign them to STOLEN case

In [106]:
df.replace("UNKNOWN", "STOLEN", inplace=True)

In [107]:
# review the target again
df.Status.value_counts()

STOLEN       25261
RECOVERED      308
Name: Status, dtype: int64

Set target to number value: STOLEN=0, RECOVERED=1

In [108]:
df.replace({'STOLEN': 0, 'RECOVERED': 1}, inplace=True)

In [109]:
df.Status.value_counts()

0    25261
1      308
Name: Status, dtype: int64

In [110]:
df.dtypes

Month      object
Div        object
Hood       object
Cost      float64
Status      int64
dtype: object

In [111]:
df['Hood'] = df['Hood'].astype(int)

In [112]:
# convert Division number part
df['Div'] = df['Div'].str.slice(1)
df['Div'] = df['Div'].astype(int)

In [113]:
df.head(5)

Month  Div  Hood    Cost  Status
0    October   22    15   700.0       0
1   November   22    15  1100.0       1
2  September   22    15   904.0       0
3        May   22    15   400.0       0
4        May   22    15   600.0       0

In [114]:
def monthN(s):
    switcher = {
        'January':      1,
        'February':     2,
        'March':        3,
        'April':        4,
        'May':          5,
        'June':         6,
        'July':         7,
        'August':       8,
        'September':    9,
        'October':      10,
        'November':     11,
        'December':     12
    }
    return switcher.get(s, "Invalid month")


In [115]:
df['Month'] = df['Month'].apply(lambda x: monthN(x))

In [116]:
df.head(5)

Month  Div  Hood    Cost  Status
0     10   22    15   700.0       0
1     11   22    15  1100.0       1
2      9   22    15   904.0       0
3      5   22    15   400.0       0
4      5   22    15   600.0       0

In [117]:
df.isnull().sum()

Month        0
Div          0
Hood         0
Cost      1744
Status       0
dtype: int64

In [118]:
df['Cost'].describe()

count     23825.000000
mean        949.542371
std        1675.880345
min           0.000000
25%         350.000000
50%         600.000000
75%        1000.000000
max      120000.000000
Name: Cost, dtype: float64

In [119]:
df['Cost'].replace({0: np.nan}, inplace=True)

In [120]:
df['Cost'].describe()

count     22657.000000
mean        998.492607
std        1704.256048
min           1.000000
25%         400.000000
50%         650.000000
75%        1051.000000
max      120000.000000
Name: Cost, dtype: float64

In [126]:
df.Cost.fillna(df["Cost"].median(), inplace = True)

In [127]:
df

Month  Div  Hood    Cost  Status
0         10   22    15   700.0       0
1         11   22    15  1100.0       1
2          9   22    15   904.0       0
3          5   22    15   400.0       0
4          5   22    15   600.0       0
...      ...  ...   ...     ...     ...
25564      4   42   132   600.0       0
25565      5   42   132   900.0       0
25566      6   42   132   700.0       0
25567      7   42   132  3000.0       0
25568      7   42   132   200.0       0

[25569 rows x 5 columns]

In [128]:
X = df[['Month', 'Div', 'Hood', 'Cost']]
y = df[['Status']]

In [130]:
X.head(5), y.head(5)

(   Month  Div  Hood    Cost
 0     10   22    15   700.0
 1     11   22    15  1100.0
 2      9   22    15   904.0
 3      5   22    15   400.0
 4      5   22    15   600.0,
    Status
 0       0
 1       1
 2       0
 3       0
 4       0)